### ISSUE TITLE
Very high cpu uses

### ISSUE THREAD

**Issue by manish1635** [Consensys-Incorporated/quorum#813]
I have implemented Quorum and observed that every interaction with Quorum led to very high CPU uses (around 90%).  Further, I went ahead and used JMETR and replay these codes 10 times concurrently and observed the same CPU uses.  
Its bit of strange behavior, I am observing in terms of CPU consumption. It shows the same CPU consumption for single execution and for 10 concurrent execution.  I have checked Memory and 
 I/O is working as expected.  I don’t know what is wrong with my implementation. Can someone help me to understand this?

var myTokenContract = web3.eth.contract(tokenContract);
var myContractInstance = myTokenContract.at(tokenAddress);
var result = myContractInstance.bestInitiativeDetails(0);

**Comment by fixanoid** [comment]
@manish1635 depending on what you do the node may need to load / update trie state and this is an io bound operation with levelDB. Is it possible to share the code thats generating high cpu load as well as your Quorum configuration: start up params, genesis, number of nodes, etc?

**Comment by manish1635** [comment]
Thanks for your reply. Required details are:

function "bestInitiativeDetails" is a simple get operation:
    function bestInitiativeDetails(uint _index) constant returns (address, uint){
        return (bestBalances[_index].initiativeAddress, bestBalances[_index].balance);
    }
However, If i don't call above method "bestInitiativeDetails" and simply runs below 2 lines even then ends up with very high CPU uses. 
var myTokenContract = web3.eth.contract(tokenContract);
var myContractInstance = myTokenContract.at(tokenAddress);

please find attached genesis file and bootstrap.sh. I am using url: https://github.com/jpmorganchase/quorum-examples to download and running 7 nodes example. Also, using privacyImpl=constellation.
In bootstrap.sh file, I did following changes:
	1. changed path from /home/vagrant/.profile to /root/.profile
	2. changed path from /home/vagrant/.bashrc to /root/.bashrc
	3. deleted 'wrk' from build deps
	4. commented copy examples
Except changes in bootstrap.sh, everything is as usual. no changes in other files after download. 

Please let me know in case more file is needed. Looking forward to hearing from your side. 


   
[Reply.zip](https://github.com/jpmorganchase/quorum/files/3588666/Reply.zip)

**Comment by fixanoid** [comment]
@manish1635 thanks for sharing these details, but I was requesting actual dapp and solidity code you are executing. I dont see it in the attachment.

**Comment by manish1635** [comment]
Due to some constraint, can't share dapp and solidity code. However, I can try to implement solutions suggested and let you know about the performance improvements. 

Also, could you please confirm that quorum installations and related settings are ok?

**Comment by fixanoid** [comment]
@manish1635 my comments:
- genesis config section should have "constantinopleBlock":0 as well
- bootstrap pulls Quorum v2.2.1, I would suggest upgrading to latest prod 2.2.5
- bootstrap: you can remove porosity install

If you share your `geth` start params, I can comment on that as well, but none of the files here should have resulted in high CPU usage.

I do recall that account unlocking and tx signing are high CPU ops in the `geth` client-- in your jmeter script, do you constantly unlock the account to execute dapp methods?

**Comment by manish1635** [comment]
Hi,
I have upgraded installation to latest prod 2.2.5 with suggested changes (genesis.json and bootstrap.sh). But, Still no luck
Also, I am not unlocking account to execute dapp methods from JMETER.

Apart from this, Please find attached "deploy.js". I am using "deploy.js" file to deploy compiled web3deploy (from remix). 
sudo geth attach http://localhost:22000
loadScript('deploy.js') 

[deploy.zip](https://github.com/jpmorganchase/quorum/files/3608379/deploy.zip)

**Comment by jbhurat** [comment]
Hi @manish1635, do you see the same issue when you are not using jmeter, like running the same transactions in a loop directly in the `geth` console

**Comment by manish1635** [comment]
Hi - Yes, same issue even do transactions in the loop. 
Apart from this, I have enhanced my hardware configuration including processor configuration (almost doubled). Overall, CPU uses don't reach 90% when iterating 10 times. However, I am still observing same spike when doing operations with Quorum.

**Comment by jbhurat** [comment]
Hi @manish1635, I have been trying to reproduce the issue, but so far no luck. As we don't have the bytecode or solidity code for the contract that you are using, I have tested by deploying https://docs.openzeppelin.com/contracts/2.x/tokens#_constructing_an_erc20_token_contract. I created the contract in loops of 100, but don't see spike in CPUs as you have mentioned. 

After running the tests I don't believe the issue is with Quorum. Is it possible to share the bytecode so that we can try to reproduce the issue.
